File: src/main/java/club/xiaojiawei/listener/log/DeckLogListener.java
package club.xiaojiawei.listener.log;

import club.xiaojiawei.bean.Deck;
import club.xiaojiawei.interfaces.closer.LogListenerCloser;
import club.xiaojiawei.data.SpringData;
import club.xiaojiawei.utils.PowerLogUtil;
import lombok.Getter;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;

import java.io.IOException;
import java.util.LinkedList;
import java.util.Objects;
import java.util.concurrent.TimeUnit;

/**
 * 套牌日志监听器
 * @author 肖嘉威
 * @date 2023/9/20 16:43
 */
@Slf4j
@Component
public class DeckLogListener extends AbstractLogListener implements LogListenerCloser {

    @Getter
    private static final LinkedList<Deck> DECKS = new LinkedList<>();

    @Autowired
    public DeckLogListener(SpringData springData) {
        super(springData.getDeckLogName(), 0, 1500L, TimeUnit.MILLISECONDS);
    }

    @Override
    protected void readOldLog() throws IOException {
        String line;
        while ((line = accessFile.readLine()) != null){
            if (line.contains("Deck Contents Received")){
                dealReceived();
            }else if (line.contains("Finished Editing Deck")){
                dealEditing();
            }
        }
    }

    private void dealReceived() throws IOException {
        DECKS.clear();
        String line;
        long filePointer = accessFile.getFilePointer();
        while ((line = accessFile.readLine()) != null){
            if (!line.contains("#")){
                accessFile.seek(filePointer);
                break;
            }
            DECKS.addFirst(createDeck(line));
            filePointer = accessFile.getFilePointer();
        }
    }

    private void dealEditing() throws IOException {
        Deck deck = createDeck(accessFile.readLine());
        boolean exist = false;
        for (Deck d : DECKS) {
            if (Objects.equals(d.getId(), deck.getId())){
                d.setName(deck.getName());
                d.setCode(deck.getCode());
                exist = true;
                break;
            }
        }
        if (!exist){
            DECKS.addFirst(deck);
        }
    }

    private Deck createDeck(String line) throws IOException {
        return new Deck(
                PowerLogUtil.iso88591_To_utf8(line.substring(line.indexOf("#") + 4)),
                (line = accessFile.readLine()).substring(line.indexOf("#") + 11),
                (line = accessFile.readLine()).substring(line.lastIndexOf(" ") + 1));
    }

    @Override
    protected void listenLog() throws IOException {
        readOldLog();
    }

    @Override
    public void closeLogListener() {
        cancelListener();
    }

}

File: Hearthstone-Script/src/main/java/club/xiaojiawei/hsscript/listener/log/DeckLogListener.kt
package club.xiaojiawei.hsscript.listener.log

import club.xiaojiawei.hsscript.bean.Deck
import club.xiaojiawei.hsscript.listener.WorkListener
import club.xiaojiawei.hsscript.status.PauseStatus
import club.xiaojiawei.hsscript.utils.PowerLogUtil
import java.util.*
import java.util.concurrent.TimeUnit

/**
 * 套牌日志监听器
 * @author 肖嘉威
 * @date 2023/9/20 16:43
 */

object DeckLogListener : AbstractLogListener("Decks.log", 0, 1500L, TimeUnit.MILLISECONDS) {

    val DECKS = LinkedList<Deck>()

    var dealing = false

    override fun dealOldLog() {
        if (dealing) return
        dealing = true
        innerLogFile?.let { file ->
            var line: String?
            while (!PauseStatus.isPause && WorkListener.working) {
                line = file.readLine()
                if (line == null || line.isEmpty()) break
                if (line.contains("Deck Contents Received")) {
                    dealReceived()
                } else if (line.contains("Finished Editing Deck")) {
                    dealEditing()
                }
            }
        }
        dealing = false
    }

    private fun dealReceived() {
        DECKS.clear()
        var line: String
        var filePointer = innerLogFile!!.filePointer
        while ((innerLogFile!!.readLine().also { line = it }) != null) {
            if (!line.contains("#")) {
                innerLogFile!!.seek(filePointer)
                break
            }
            DECKS.addFirst(createDeck(line))
            filePointer = innerLogFile!!.filePointer
        }
    }

    private fun dealEditing() {
        val deck = createDeck(innerLogFile!!.readLine())
        var exist = false
        for (d in DECKS) {
            if (d.id == deck.id) {
                d.apply {
                    name = deck.name
                    code = deck.code
                }
                exist = true
                break
            }
        }
        if (!exist) {
            DECKS.addFirst(deck)
        }
    }

    private fun createDeck(line: String): Deck {
        var l = line
        return Deck(
            PowerLogUtil.iso88591ToUtf8(l.substring(l.indexOf("#") + 4)),
            (innerLogFile!!.readLine().also { l = it }).substring(l.indexOf("#") + 11),
            (innerLogFile!!.readLine().also { l = it }).substring(l.lastIndexOf(" ") + 1)
        )
    }

    override fun dealNewLog() {
        dealOldLog()
    }

}
